### ISSUE TITLE
GRPO training for caption prediction evaluated using BELU score

### ISSUE THREAD

**Issue by yhifny** [2U1/Qwen2-VL-Finetune#160]
Hi, I would like to train Qwen 2-VL-2B using GRPO using a reward function (BELU or ROUGEL). Do you know how I can do that? Many thanks for considering my request.

**Comment by 2U1** [comment]
You could make a reward function like this
```
from sacrebleu import corpus_bleu
from rouge_score import rouge_scorer

_bleu = lambda cand, ref: corpus_bleu(cand, [ref]).score / 100.0 
_rouge = rouge_scorer.RougeScorer(["rougeL"], use_stemmer=True)

def bleu_reward(completions, assistant, **kwargs):
    cands  = [c[0]["content"] for c in completions]
    refs   = [a["content"]    for a in assistant]
    return [_bleu(c, r) for c, r in zip(cands, refs)]

def rougel_reward(completions, assistant, **kwargs):
    cands = [c[0]["content"] for c in completions]
    refs  = [a["content"]    for a in assistant]
    return [_rouge.score(r, c)["rougeL"].fmeasure for c, r in zip(cands, refs)]

```
You need to scale the score between 0~1 to traine the model stable.

**Comment by yhifny** [comment]
Thanks a lot. now what is the suitable format for the training dataset:

**Comment by yhifny** [comment]
Can you please give an example for the training sample JSON dictionary? I mean how to set `"<think>? </think>" and "<answer>?</answer>" `

**Comment by yhifny** [comment]
What do u mean "You need to scale the score between 0~1 to train the model stable."? I found that roughl and belu scores are normalized between 0 and 1 based on the code that you wrote. So, I do not need to normalize these scores, right?

**Comment by yhifny** [comment]
I have 10,000 samples, and the batch size is 4, so I have 2,500 iterations for an epoch. Then, three epochs lead to 7,500 iterations. However, I found in the log file of the GRPO training that it is "0%|          | 1/15000", so it should be  7,500 but it is actually 15,000. Can you please clarify the reason for this mismatch?

**Comment by 2U1** [comment]
https://github.com/2U1/Qwen2-VL-Finetune/issues/154#issuecomment-3002261699

This could be the answer for the question.

**Comment by yhifny** [comment]
The loss is zero after 500 steps. Do you think it makes sense? "{'loss': 0.0, 'grad_norm': 0.0, 'learning_rate': 4.990231533628719e-06, 'num_tokens': 6523696.0, 'completions/mean_length':
124.0, 'completions/min_length': 115.0, 'completions/max_length': 133.0, 'completions/clipped_ratio': 0.0, 'completions/mean_terminated_length': 124.0, 'completions/min_terminated_length': 115.0, 'completions/max_terminated_length': 133.0, 'rewards/accuracy_reward/mean': 0.5, 'rewards/accuracy_reward/std': 0.5773502588272095, 'rewards/format_reward/mean': 0.0, 'rewards/format_reward/std': 0.0, 'rewards/bleu_reward/mean': 0.0, 'rewards/bleu_reward/std': 0.0, 'rewards/rougel_reward/mean': 0.346230685710907, 'rewards/rougel_reward/std': 0.005365344695746899, 'reward': 0.846230685710907, 'reward_std': 0.0, 'kl': 0.0, 'clip_ratio/low_mean': 0.0, 'clip_ratio/low_min': 0.0, 'clip_ratio/high_mean': 0.0, 'clip_ratio/high_max': 0.0, 'clip_ratio/region_mean': 0.0, 'epoch': 0.11}"

**Comment by 2U1** [comment]
Maybe it could be a problem with
1. reward score,
2. Non EOS generating 
3. Deteministic sampling 

You could add the debug script
```
    @profiling_decorator
    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        if return_outputs:
            raise ValueError("The GRPOTrainer does not support returning outputs")

        # Check if we need to generate new completions or use buffered ones
        if self.state.global_step % self.num_iterations == 0:
            inputs = self._generate_and_score_completions(inputs)
            self._buffered_inputs[self._step % self.args.gradient_accumulation_steps] = inputs
        else:
            inputs = self._buffered_inputs[self._step % self.args.gradient_accumulation_steps]
        self._step += 1

        print(
            f"step {self.state.global_step}  "
            f"adv_max={inputs['advantages'].abs().max().item():.4f}  "
            f"trainable={sum(p.requires_grad for p in self.model.parameters())}"
        )

        if self.use_liger_loss:
            # Compute the loss using the liger grpo loss
            return self.compute_liger_loss(model, inputs)
        else:
            return self._compute_loss(model, inputs)
``` 
something like this and check it.

**Comment by yhifny** [comment]
`{'loss': 0.0, 'grad_norm': 0.0, 'learning_rate': 4.985462964079137e-06, 'num_tokens': 7407600.0, 'completions/mean_length': 98.0, 'completions/min_length': 98.0, 'completions/max_length': 98.0, 'completions/clipped_ratio': 0.0, 'completions/mean_terminated_length':
98.0, 'completions/min_terminated_length': 98.0, 'completions/max_terminated_length': 98.0, 'rewards/bleu_reward/mean': 0.0, 'rewards/bleu_reward/std': 0.0, 'rewards/rougel_reward/mean': 0.310502290725708, 'rewards/rougel_reward/std': 0.026363028213381767, 'reward': 0.310502290725708, 'reward_std': 0.0, 'kl': 0.0, 'clip_ratio/low_mean': 0.0, 'clip_ratio/low_min': 0.0, 'clip_ratio/high_mean': 0.0, 'clip_ratio/high_max': 0.0, 'clip_ratio/region_mean': 0.0, 'epoch': 0.13}

  4%|▍         | 661/15000 [7:37:46<617:49:35, 155.11s/it]step 661  adv_max=0.0000  trainable=392
step 661  adv_max=0.0000  trainable=392
step 661  adv_max=0.0000  trainable=392
step 661  adv_max=0.0000  trainable=392` **The values of adv_max and trainable are the same for each update**?

**Comment by yhifny** [comment]
I set `    --num_generations 4 \
    --max_completion_length 1024 \
    --max_prompt_length 512 \
    --temperature   1.1\
    --top_p 1.0\
    --top_k 50 \
    --repetition_penalty 1.0 \` 

 But I see this message in the log:

> `generation_config` default values have been modified to match model-specific defaults: {'top_k': 1, 'top_p': 0.001, 'bos_token_id':151643}. If this is not desired, please set these values explicitly."

How do you think I could fix this issue?

**Comment by 2U1** [comment]
I've made some update for the generation config.
I think it wasn't properly applied. You could just copy & paste the grpo_trainer code and retry.
